sycl/doc/SYCLPluginInterface.md

```diff
@@ -41,12 +40,45 @@ once before any actual offload is attempted.
 
 ### Plugin discovery
 
-Plugins are physically dynamic libraries stored somewhere in the system where
-the SYCL runtime runs. TBD - design and describe the process in details.
+Plugins are physically dynamic libraries or shared objects.
+The process to discover plugins will follow the following guidelines.
+
+The SYCL Runtime will search for plugins at env LD_LIBRARY_PATH location on
+Linux and env LIB on Windows, with names in the format of libpiXXX.so and
+query them. An extension to the search mechanism is to use a configuration file
+which lists all the available plugins and their locations. 
+This file can be stored along with SYCL Runtime.
+Plugins should expose the information of supported PI API version as a constant
+string, so it can be read without loading the library.
```

Comment:
I thought that a sufficiently complex mangled name would do the trick, e.g.
~~~
const char *__libsyclpiopencl_so_version__ = "__libsyclpiopencl_so_version__x.y.z"
const char *__libsyclpiopencl_pi_version__ = "__libsyclpiopencl_pi_version__x.y.z"
~~~
This can be searched in any OS with tools supporting binary object dumps.
On top we should still probably have an entry that will just return/dump the version.

Reply:
Could you share even more information of how you going to check symbol without loading library?
And what is the reason for avoiding loading of library?

Reply:
Will there be different variants of the same plugin? e.g., two different implementations of the piopencl plugin ?

Reply:
~~~
$ strings libsyclpiopencl.so | grep __libsyclpiopencl_so_version__.
__libsyclpiopencl_so_version__x.y.z
~~~

> Will there be different variants of the same plugin? e.g., two different implementations of the piopencl plugin ?

I don't see why this would not be possible, or that we should specially regulate that.

Reply:
> And what is the reason for avoiding loading of library?

This way an installation can be verified (e.g. against compatibility of compatibility of contained components) without needing to actually run anything.

Reply:
> And what is the reason for avoiding loading of library?

one of the reasons is security - not running static initialization code of (in general case) unknown library.

Reply:
> > Will there be different variants of the same plugin? e.g., two different implementations of the piopencl plugin ?
> 
> I don't see why this would not be possible, or that we should specially regulate that.

Is that an user selection based on the path? , e.g. point LD_LIBRARY_PATH to my desired piocl implementation,
or the SYCL runtime see all the variants of piocl and decides which one to use?

Reply:
> > And what is the reason for avoiding loading of library?
> 
> one of the reasons is security - not running static initialization code of (in general case) unknown library.

Sure, but if this is really the problem, there are quite many more security flaws in the approach... :-)

Reply:
> Is that an user selection based on the path? , e.g. point LD_LIBRARY_PATH to my desired piocl implementation,
> the SYCL runtime see all the variants of piocl and decides which one to use?

Both are OK. The LD_LIBRARY_PATH and/or configuration file will enable the discovery of the plugins, and SYCL RT would choose which one to use based on its actual heuristics. On top there would be a user control (environment variable) to influence SYCL RT decision.

Reply:
Currently the support is added to check for the plugin in LD_LIBRARY_PATH.
In review https://github.com/intel/llvm/pull/708.

Reply:
@garimagu , my impression was that we agreed that SYCL plugin loader does not search for libpi* files in LD_LIBRARY_PATH, but takes the names of the plugin files from the configuration file which must be present. LD_LIBRARY_PATH is still used as a search path for these, though. If you agree - please update the doc.

There were couple more suggestions, which I think can be implemented later atop:
> It would be good if it would not be necessary to add plugin directories to LD_LIBRARY_PATH (on Linux, nor PATH on Windows) for the runtime to use the plugins packaged with it.

this one maybe implemented either as an absolute path in the config file or as searching in some predefined directory (like <dirname of sycl.so>/plugins)

And the other one:
> On Linux an DT_RUNPATH can be embedded into the SYCL library for dlopen to look relative to the location of the SYCL library for plugin libraries

There should also be environment variables overriding any of the above.

Reply:
Doc updated to be consistent with the last comment by Kbobrovs.